File: python/navigation/local_planning/reactive/dwa.py
"""Dynamic Window Approach (Fox, Burgard & Thrun 1997, DOI 10.1109/100.580977):
searches the reachable (v, omega) command space directly instead of a
Cartesian path. Every tick samples a deterministic grid inside the box formed
by the robot's kinematic limits intersected with what one accel-limited tick
can reach from the current velocity, rolls each candidate forward as a
constant-command arc, discards ones that collide or cannot stop before the
nearest obstacle, and picks the highest-scoring survivor.
"""

from __future__ import annotations

import math

from navigation.core.capabilities import Capability, ObstacleQuery
from navigation.core.params import ParamSet
from navigation.core.planner import ObstacleLocalPlanner
from navigation.core.trace import TraceRecorder
from navigation.core.types import Footprint, LocalTask, Point, Pose, RobotState, VelocityCommand

from .._geometry import wrap_to_pi

# Below this |omega| the closed-form arc (division by omega) is numerically
# unstable, so the rollout falls back to the straight-line limit instead --
# the same threshold and reasoning as simulation.py's _OMEGA_EPS. Duplicated
# rather than imported: algorithm modules depend on core only, never on the
# simulator (see local_planning/simulation.py's own docstring).
_OMEGA_EPS = 1e-9

class Dwa(ObstacleLocalPlanner):
    def __init__(self, params: ParamSet) -> None:
        super().__init__(params)
        self._max_speed = params.get_float("max_speed")
        self._min_speed = params.get_float("min_speed")
        self._max_omega = params.get_float("max_omega")
        self._accel = params.get_float("accel")
        self._accel_omega = params.get_float("accel_omega")
        self._v_samples = params.get_int("v_samples")
        self._omega_samples = params.get_int("omega_samples")
        self._sim_time = params.get_float("sim_time")
        self._sim_steps = params.get_int("sim_steps")
        self._heading_weight = params.get_float("heading_weight")
        self._clearance_weight = params.get_float("clearance_weight")
        self._velocity_weight = params.get_float("velocity_weight")
        self._clearance_limit = params.get_float("clearance_limit")
        self._slow_radius = params.get_float("slow_radius")
        self._footprint_radius = params.get_float("footprint_radius")

    @property
    def name(self) -> str:
        return "dwa"

    def required_capabilities(self) -> set[Capability]:
        return {Capability.OBSTACLE_QUERY}

    def _rollout(self, pose: Pose, v: float, omega: float) -> list[Pose]:
        # Each candidate is scored by holding (v, omega) constant for sim_time,
        # sampled at sim_steps equally spaced instants (Fox 1997's circular-arc
        # trajectory prediction) -- every point is computed directly from the
        # start pose (not chained step-to-step), which is exact for a constant
        # command and matches simulation.py's per-tick integrate_unicycle.
        x, y, theta = pose
        poses: list[Pose] = []
        for k in range(1, self._sim_steps + 1):
            t = self._sim_time * k / self._sim_steps
            if abs(omega) < _OMEGA_EPS:
                poses.append((x + v * t * math.cos(theta), y + v * t * math.sin(theta), theta))
                continue
            new_theta = theta + omega * t
            px = x + (v / omega) * (math.sin(new_theta) - math.sin(theta))
            py = y - (v / omega) * (math.cos(new_theta) - math.cos(theta))
            poses.append((px, py, wrap_to_pi(new_theta)))
        return poses

    def _score(
        self,
        space: ObstacleQuery,
        footprint: Footprint,
        rollout: list[Pose],
        v: float,
        omega: float,
        goal_x: float,
        goal_y: float,
    ) -> tuple[float, float, float, float, bool]:
        """Returns (cost, heading, clearance, velocity, admissible). A colliding
        rollout is rejected outright (cost=0, the three terms zeroed) rather
        than scored, per Fox 1997's hard obstacle constraint."""
        if any(space.is_collision(footprint, pose) for pose in rollout):
            return 0.0, 0.0, 0.0, 0.0, False

        # Fox 1997 eq. 14 uses the curvature's true distance to the nearest
        # obstacle; this implementation approximates it conservatively with the
        # minimum clearance sampled along the finite rollout.
        nearest = min(space.distance_to_nearest((px, py)) for px, py, _ in rollout)
        clearance = max(0.0, nearest - self._footprint_radius)
        x_end, y_end, theta_end = rollout[-1]
        # Fox 1997 eq. 15's target-heading term, evaluated at this
        # implementation's rollout endpoint rather than the paper's
        # maximum-deceleration stopping position (a deliberate simplification).
        goal_bearing = math.atan2(goal_y - y_end, goal_x - x_end)
        heading = 1.0 - abs(wrap_to_pi(goal_bearing - theta_end)) / math.pi
        velocity = v / self._max_speed
        clearance_term = min(clearance, self._clearance_limit) / self._clearance_limit
        # Fixed (non-batch) normalization: scoring one candidate never depends
        # on the rest of the candidate set, which keeps py/cpp/TS scoring
        # bit-identical and reproducible test-to-test, unlike Fox 1997's
        # per-tick min-max smoothing over the whole candidate batch.
        cost = (
            self._heading_weight * heading
            + self._clearance_weight * clearance_term
            + self._velocity_weight * velocity
        )
        # Fox 1997 eq. 14: admissible iff the robot could stop (at accel/
        # accel_omega) before covering `clearance` -- a stopping-distance bound,
        # approximated here with the finite-rollout clearance above.
        admissible = v <= math.sqrt(2.0 * clearance * self._accel) and abs(omega) <= math.sqrt(
            2.0 * clearance * self._accel_omega
        )
        return cost, heading, clearance, velocity, admissible

    def _decelerate(self, v_a: float, omega_a: float, dt: float) -> tuple[float, float]:
        # No admissible candidate survived this tick (e.g. boxed in on every
        # side): brake at the kinematic limits rather than execute an
        # unscored guess. Persistent triggering here is a local minimum, which
        # the simulator's stall detector reports honestly as STALLED.
        v_cmd = max(0.0, v_a - self._accel * dt)
        sign = 1.0 if omega_a > 0.0 else (-1.0 if omega_a < 0.0 else 0.0)
        omega_cmd = omega_a - sign * min(abs(omega_a), self._accel_omega * dt)
        return v_cmd, omega_cmd

    def compute_command(
        self,
        space: ObstacleQuery,
        state: RobotState,
        task: LocalTask,
        dt: float,
        recorder: TraceRecorder | None = None,
    ) -> VelocityCommand:
        x, y, theta = state.pose
        v_a, omega_a = state.v, state.omega
        goal_x, goal_y = task.goal[0], task.goal[1]

        remaining = math.hypot(goal_x - x, goal_y - y)
        # Goal-proximity speed cap: not part of Fox 1997, a practical extension
        # so an episode ends in REACHED instead of orbiting/overshooting the
        # goal at max_speed.
        v_max_eff = self._max_speed * min(1.0, remaining / self._slow_radius)

        v_lo = max(self._min_speed, v_a - self._accel * dt)
        v_hi = min(v_max_eff, v_a + self._accel * dt)
        omega_lo = max(-self._max_omega, omega_a - self._accel_omega * dt)
        omega_hi = min(self._max_omega, omega_a + self._accel_omega * dt)

        best_index: int | None = None
        best_cost = 0.0
        best_v, best_omega = 0.0, 0.0
        # Trace buffering: only populated when a recorder is attached (zero
        # cost otherwise) because `selected` cannot be known until every
        # candidate in the tick has been scored.
        buffered: list[tuple[Point, float, dict[str, float], list[Point]]] = []

        if v_lo <= v_hi and omega_lo <= omega_hi:
            footprint = Footprint(self._footprint_radius)
            v_step = (v_hi - v_lo) / (self._v_samples - 1) if self._v_samples > 1 else 0.0
            n_omega = self._omega_samples
            omega_step = (omega_hi - omega_lo) / (n_omega - 1) if n_omega > 1 else 0.0
            candidate_index = 0
            # Deterministic uniform grid (never random sampling): fixed
            # traversal order (v outer, omega inner) keeps py/cpp/TS scoring
            # and tie-breaking bit-identical.
            for i in range(self._v_samples):
                v = v_lo + v_step * i
                for j in range(self._omega_samples):
                    omega = omega_lo + omega_step * j
                    rollout = self._rollout((x, y, theta), v, omega)
                    cost, heading, clearance, velocity, admissible = self._score(
                        space, footprint, rollout, v, omega, goal_x, goal_y
                    )
                    if admissible and (best_index is None or cost > best_cost):
                        best_index, best_cost, best_v, best_omega = candidate_index, cost, v, omega
                    if recorder is not None:
                        x_end, y_end, _ = rollout[-1]
                        data = {
                            "v": v,
                            "omega": omega,
                            "heading": heading,
                            "clearance": clearance,
                            "velocity": velocity,
                            "admissible": 1.0 if admissible else 0.0,
                            "selected": 0.0,  # finalized below once selection is known
                        }
                        rollout_xy = [(p[0], p[1]) for p in rollout]
                        buffered.append(((x_end, y_end), cost, data, rollout_xy))
                    candidate_index += 1

        if best_index is not None:
            v_cmd, omega_cmd = best_v, best_omega
        else:
            v_cmd, omega_cmd = self._decelerate(v_a, omega_a, dt)

        if recorder is not None:
            for idx, (state_xy, cost, data, rollout_xy) in enumerate(buffered):
                data["selected"] = 1.0 if idx == best_index else 0.0
                recorder.candidate_evaluated(state_xy, cost, data=data, rollout=rollout_xy)

        return VelocityCommand(v_cmd, omega_cmd)

File: cpp/src/local_planning/reactive/dwa.cpp
#include "navigation/local_planning/reactive/dwa.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <map>
#include <string>
#include <tuple>
#include <utility>
#include <vector>

#include "navigation/local_planning/geometry.hpp"

namespace navigation::local_planning {

namespace {

// Below this |omega| the closed-form arc (division by omega) is numerically
// unstable, so the rollout falls back to the straight-line limit instead --
// the same threshold and reasoning as simulation.cpp's inline 1e-9 guard.
// Duplicated rather than shared: algorithm modules depend on core only, never
// on the simulator.
constexpr double kOmegaEps = 1e-9;

}  // namespace

DwaPlanner::DwaPlanner(core::ParamSet params)
    : core::ObstacleLocalPlanner(std::move(params)),
      max_speed_(params_.get_float("max_speed")),
      min_speed_(params_.get_float("min_speed")),
      max_omega_(params_.get_float("max_omega")),
      accel_(params_.get_float("accel")),
      accel_omega_(params_.get_float("accel_omega")),
      v_samples_(params_.get_int("v_samples")),
      omega_samples_(params_.get_int("omega_samples")),
      sim_time_(params_.get_float("sim_time")),
      sim_steps_(params_.get_int("sim_steps")),
      heading_weight_(params_.get_float("heading_weight")),
      clearance_weight_(params_.get_float("clearance_weight")),
      velocity_weight_(params_.get_float("velocity_weight")),
      clearance_limit_(params_.get_float("clearance_limit")),
      slow_radius_(params_.get_float("slow_radius")),
      footprint_radius_(params_.get_float("footprint_radius")) {}

void DwaPlanner::rollout(const core::Pose& pose, double v, double omega,
                         std::vector<core::Pose>& out) const {
  // Each candidate is scored by holding (v, omega) constant for sim_time,
  // sampled at sim_steps equally spaced instants (Fox 1997's circular-arc
  // trajectory prediction) -- every point is computed directly from the start
  // pose (not chained step-to-step), which is exact for a constant command
  // and matches simulation.cpp's per-tick integrate_unicycle.
  const double x = pose.x, y = pose.y, theta = pose.theta;
  out.clear();  // keeps capacity: no per-candidate allocation in the hot loop
  for (int k = 1; k <= sim_steps_; ++k) {
    const double t = sim_time_ * static_cast<double>(k) / static_cast<double>(sim_steps_);
    if (std::fabs(omega) < kOmegaEps) {
      out.push_back(core::Pose{x + v * t * std::cos(theta), y + v * t * std::sin(theta), theta});
      continue;
    }
    const double new_theta = theta + omega * t;
    const double px = x + (v / omega) * (std::sin(new_theta) - std::sin(theta));
    const double py = y - (v / omega) * (std::cos(new_theta) - std::cos(theta));
    out.push_back(core::Pose{px, py, wrap_to_pi(new_theta)});
  }
}

DwaPlanner::Score DwaPlanner::score(core::ObstacleQuery& space, const core::Footprint& footprint,
                                    const std::vector<core::Pose>& rollout, double v, double omega,
                                    double goal_x, double goal_y) const {
  for (const core::Pose& pose : rollout) {
    if (space.is_collision(footprint, pose)) return Score{};
  }

  // Fox 1997 eq. 14 uses the curvature's true distance to the nearest
  // obstacle; this implementation approximates it conservatively with the
  // minimum clearance sampled along the finite rollout.
  double nearest = std::numeric_limits<double>::infinity();
  for (const core::Pose& pose : rollout) {
    nearest = std::min(nearest, space.distance_to_nearest(core::Point{pose.x, pose.y}));
  }
  const double clearance = std::max(0.0, nearest - footprint_radius_);
  const core::Pose& end = rollout.back();
  // Fox 1997 eq. 15's target-heading term, evaluated at this implementation's
  // rollout endpoint rather than the paper's maximum-deceleration stopping
  // position (a deliberate simplification).
  const double goal_bearing = std::atan2(goal_y - end.y, goal_x - end.x);
  const double heading = 1.0 - std::fabs(wrap_to_pi(goal_bearing - end.theta)) / M_PI;
  const double velocity = v / max_speed_;
  const double clearance_term = std::min(clearance, clearance_limit_) / clearance_limit_;

  Score result;
  result.heading = heading;
  result.clearance = clearance;
  result.velocity = velocity;
  // Fixed (non-batch) normalization: scoring one candidate never depends on
  // the rest of the candidate set, which keeps py/cpp/TS scoring
  // bit-identical and reproducible test-to-test, unlike Fox 1997's per-tick
  // min-max smoothing over the whole candidate batch.
  result.cost = heading_weight_ * heading + clearance_weight_ * clearance_term + velocity_weight_ * velocity;
  // Fox 1997 eq. 14: admissible iff the robot could stop (at accel/
  // accel_omega) before covering `clearance` -- a stopping-distance bound,
  // approximated here with the finite-rollout clearance above.
  result.admissible = v <= std::sqrt(2.0 * clearance * accel_) &&
                      std::fabs(omega) <= std::sqrt(2.0 * clearance * accel_omega_);
  return result;
}

std::pair<double, double> DwaPlanner::decelerate(double v_a, double omega_a, double dt) const {
  const double v_cmd = std::max(0.0, v_a - accel_ * dt);
  const double sign = omega_a > 0.0 ? 1.0 : (omega_a < 0.0 ? -1.0 : 0.0);
  const double omega_cmd = omega_a - sign * std::min(std::fabs(omega_a), accel_omega_ * dt);
  return {v_cmd, omega_cmd};
}

core::VelocityCommand DwaPlanner::compute_command(core::ObstacleQuery& space,
                                                   const core::RobotState& state,
                                                   const core::LocalTask& task, double dt,
                                                   core::TraceRecorder* recorder) {
  const double x = state.pose.x, y = state.pose.y, theta = state.pose.theta;
  const double v_a = state.v, omega_a = state.omega;
  const double goal_x = task.goal.x, goal_y = task.goal.y;

  const double remaining = std::hypot(goal_x - x, goal_y - y);
  // Goal-proximity speed cap: not part of Fox 1997, a practical extension so
  // an episode ends in REACHED instead of orbiting/overshooting the goal at
  // max_speed.
  const double v_max_eff = max_speed_ * std::min(1.0, remaining / slow_radius_);

  const double v_lo = std::max(min_speed_, v_a - accel_ * dt);
  const double v_hi = std::min(v_max_eff, v_a + accel_ * dt);
  const double omega_lo = std::max(-max_omega_, omega_a - accel_omega_ * dt);
  const double omega_hi = std::min(max_omega_, omega_a + accel_omega_ * dt);

  bool have_best = false;
  double best_cost = 0.0, best_v = 0.0, best_omega = 0.0;
  int best_index = -1;

  // Trace buffering: only populated when a recorder is attached (zero cost
  // otherwise) because `selected` cannot be known until every candidate in
  // the tick has been scored.
  struct Buffered {
    core::Point state_xy;
    double cost;
    std::map<std::string, double> data;
    std::vector<std::vector<double>> rollout_xy;
  };
  std::vector<Buffered> buffered;

  if (v_lo <= v_hi && omega_lo <= omega_hi) {
    const core::Footprint footprint{footprint_radius_};
    const double v_step = v_samples_ > 1 ? (v_hi - v_lo) / (v_samples_ - 1) : 0.0;
    const double omega_step = omega_samples_ > 1 ? (omega_hi - omega_lo) / (omega_samples_ - 1) : 0.0;
    int candidate_index = 0;
    // One rollout buffer reused across every candidate in the tick (clear()
    // keeps capacity), so the hot loop performs no per-candidate allocation.
    std::vector<core::Pose> candidate_rollout;
    candidate_rollout.reserve(static_cast<size_t>(sim_steps_));
    // Deterministic uniform grid (never random sampling): fixed traversal
    // order (v outer, omega inner) keeps py/cpp/TS scoring and tie-breaking
    // bit-identical.
    for (int i = 0; i < v_samples_; ++i) {
      const double v = v_lo + v_step * static_cast<double>(i);
      for (int j = 0; j < omega_samples_; ++j) {
        const double omega = omega_lo + omega_step * static_cast<double>(j);
        rollout(core::Pose{x, y, theta}, v, omega, candidate_rollout);
        Score s = score(space, footprint, candidate_rollout, v, omega, goal_x, goal_y);
        if (s.admissible && (!have_best || s.cost > best_cost)) {
          have_best = true;
          best_index = candidate_index;
          best_cost = s.cost;
          best_v = v;
          best_omega = omega;
        }
        if (recorder) {
          const core::Pose& end = candidate_rollout.back();
          std::vector<std::vector<double>> rollout_xy;
          rollout_xy.reserve(candidate_rollout.size());
          for (const core::Pose& pose : candidate_rollout) rollout_xy.push_back({pose.x, pose.y});
          buffered.push_back(Buffered{
              core::Point{end.x, end.y},
              s.cost,
              {{"v", v},
               {"omega", omega},
               {"heading", s.heading},
               {"clearance", s.clearance},
               {"velocity", s.velocity},
               {"admissible", s.admissible ? 1.0 : 0.0},
               {"selected", 0.0}},  // finalized below once selection is known
              std::move(rollout_xy)});
        }
        ++candidate_index;
      }
    }
  }

  double v_cmd, omega_cmd;
  if (have_best) {
    v_cmd = best_v;
    omega_cmd = best_omega;
  } else {
    std::tie(v_cmd, omega_cmd) = decelerate(v_a, omega_a, dt);
  }

  if (recorder) {
    for (int idx = 0; idx < static_cast<int>(buffered.size()); ++idx) {
      Buffered& b = buffered[static_cast<size_t>(idx)];
      b.data["selected"] = idx == best_index ? 1.0 : 0.0;
      recorder->candidate_evaluated(b.state_xy, b.cost, b.data, b.rollout_xy);
    }
  }

  return core::VelocityCommand{v_cmd, omega_cmd};
}

}  // namespace navigation::local_planning
